src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp

```diff
@@ -110,33 +110,61 @@ bool ShenandoahAdaptiveHeuristics::should_start_gc() const {
   size_t soft_tail = max_capacity - capacity;
   available = (available > soft_tail) ? (available - soft_tail) : 0;
 
+  if (is_available_below_min_threshold(capacity, available)) {
+    return true;
+  }
+
+  if (is_learning_necessary(capacity, available)) {
+    return true;
+  }
+
+  if (is_allocation_headroom_low(capacity, available)) {
+    return true;
+  }
+
+  return ShenandoahHeuristics::should_start_gc();
+}
+
+bool ShenandoahAdaptiveHeuristics::is_available_below_min_threshold(size_t capacity,
+                                                                    size_t available) const {
   // Check if we are falling below the worst limit, time to trigger the GC, regardless of
   // anything else.
   size_t min_threshold = capacity / 100 * ShenandoahMinFreeThreshold;
   if (available < min_threshold) {
-    log_info(gc)("Trigger: Free (" SIZE_FORMAT "%s) is below minimum threshold (" SIZE_FORMAT "%s)",
-                 byte_size_in_proper_unit(available),     proper_unit_for_byte_size(available),
-                 byte_size_in_proper_unit(min_threshold), proper_unit_for_byte_size(min_threshold));
+    if (log_is_enabled(Info, gc)) {
+      log_info(gc)("Trigger: Free (" SIZE_FORMAT "%s) is below minimum threshold (" SIZE_FORMAT "%s)",
```

Comment:
Why do we need these additional checks? Is this to avoid computing the arguments?  I was under impression `log_info` and friends do this right already. Are they?

Reply:
You're right, the `log_info` macro performs this check before evaluating the arguments. I'll fix this.

Reply:
Good! (please click "Resolve conversation")